attestation-agent/kbs_protocol/src/keypair.rs

```diff
@@ -3,80 +3,160 @@
+                let k_exp = URL_SAFE_NO_PAD.encode(key.e());
 
-        Ok(TeePubKey::RSA {
-            alg: PaddingMode::PKCS1v15.as_ref().to_string(),
-            k_mod,
-            k_exp,
-        })
+                Ok(TeePubKey::RSA {
+                    alg: PaddingMode::OAEP.as_ref().to_string(),
+                    k_mod,
+                    k_exp,
+                })
+            }
+            TeeKey::Ec(key) => {
+                let x = URL_SAFE_NO_PAD.encode(key.x()?);
+                let y = URL_SAFE_NO_PAD.encode(key.y()?);
+
+                Ok(TeePubKey::EC {
+                    crv: Curve::P256.as_ref().to_string(),
+                    alg: KeyWrapAlgorithm::EcdhEsA256Kw.as_ref().to_string(),
+                    x,
+                    y,
+                })
+            }
+        }
     }
 
     #[inline]
-    pub fn decrypt(&self, mode: PaddingMode, cipher_text: Vec<u8>) -> Result<Vec<u8>> {
-        self.keypair.decrypt(mode, cipher_text)
+    pub fn unwrap_cek(&self, header: &ProtectedHeader, cipher_text: Vec<u8>) -> Result<Vec<u8>> {
+        #[allow(deprecated)]
+        if &header.alg[..] == PaddingMode::PKCS1v15.as_ref() {
+            warn!("Use deprecated Rsa PKCSv1.5 algorithm!");
+            let TeeKey::Rsa(key) = &self.key else {
+                bail!("Unmatched key. Must be RSA key");
+            };
+
+            let cek = key.decrypt(PaddingMode::OAEP, cipher_text)?;
+            Ok(cek)
+        } else if &header.alg[..] == PaddingMode::OAEP.as_ref() {
+            let TeeKey::Rsa(key) = &self.key else {
+                bail!("Unmatched key. Must be RSA key");
+            };
+
+            let cek = key.decrypt(PaddingMode::OAEP, cipher_text)?;
+            Ok(cek)
+        } else if &header.alg[..] == KeyWrapAlgorithm::EcdhEsA256Kw.as_ref() {
+            let epk = header
+                .other_fields
+                .get("epk")
+                .ok_or(anyhow!("Invalid JWE ProtectedHeader. Without `epk`"))?;
+            let crv = epk
+                .get("crv")
+                .ok_or(anyhow!("Invalid JWE ProtectedHeader. Without `crv`"))?
+                .as_str()
+                .ok_or(anyhow!(
+                    "Invalid JWE ProtectedHeader. `crv` is not a string"
+                ))?;
+
+            let x = epk
+                .get("x")
+                .ok_or(anyhow!("Invalid JWE ProtectedHeader. Without `x`"))?
+                .as_str()
+                .ok_or(anyhow!("Invalid JWE ProtectedHeader. `x` is not a string"))?;
+
+            let x = URL_SAFE_NO_PAD.decode(x)?;
+
+            let y = epk
+                .get("y")
+                .ok_or(anyhow!("Invalid JWE ProtectedHeader. Without `y`"))?
+                .as_str()
+                .ok_or(anyhow!("Invalid JWE ProtectedHeader. `y` is not a string"))?;
+
+            let y = URL_SAFE_NO_PAD.decode(y)?;
+
+            let TeeKey::Ec(key) = &self.key else {
+                bail!("Unmatched key. Must be RSA key");
```

Comment:
This should probably be "Must be EC Key"

Reply:
Thanks. Let me fix